### PR TITLE
[FLINK-16579][table] Upgrade Calcite version to 1.26 for Flink SQL

### PR DESCRIPTION
**Pull request by danny0405**
## What is the purpose of the change

Upgrade the Calcite version to 1.26.0. As a dependency, the Guava version to 29.0-jre and janino version to 3.0.11

## Brief change log

  - Upgrade the sql parser calcite version to 1.26.0
  - Upgrade the planners calcite version to 1.26.0

## Verifying this change

Existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-table/flink-table-planner-blink/pom.xml [hunk 1]
```diff
@@ -189,10 +194,6 @@ under the License.
 					<groupId>org.apache.commons</groupId>
 					<artifactId>commons-dbcp2</artifactId>
 				</exclusion>
-				<exclusion>
-					<groupId>com.esri.geometry</groupId>
-					<artifactId>esri-geometry-api</artifactId>
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/pom.xml:194]: Why we have to depend on this dependency now?

File: flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java [hunk 2]
```diff
@@ -35,15 +35,12 @@ protected SqlParserImplFactory parserImplFactory() {
 	}
 
 	// overrides test methods that we don't support
-	@Override
```
Comment (wuchong) [inline flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java:38]: What about changing to use `@Ignore` +	`@Test` annotations and also update the comment above?
This can avoid the IDEA warns this method is never used.

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 3]
```diff
@@ -994,7 +992,6 @@ public void testShowViews() {
 
 	// Override the test because our ROW field type default is nullable,
 	// which is different with Calcite.
-	@Override
```
Comment (wuchong) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:997]: Use `@Test` instead?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala [hunk 4]
```diff
@@ -65,15 +65,8 @@ class TableSourceTable(
     statistic) {
 
   override def getQualifiedName: util.List[String] = {
-    val names = super.getQualifiedName
     val builder = ImmutableList.builder[String]()
-    builder.addAll(names)
-    if (dynamicOptions.size() != 0) {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala:71]: It seems that `dynamicOptions` is never used in this class, should we remove it?

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/schema/Statistic.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Statistics about a {@link Table}.
+ *
+ * <p>Each of the methods may return {@code null} meaning "not known".</p>
+ *
+ * <p>Changes:
+ *
+ * <ul>
+ *     <li>Line 61: default collations change from null to empty list.</li>
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/main/java/org/apache/calcite/schema/Statistic.java:35]: Is it a temporary solution? Why we have to override the default value in interface?
But this change is only needed in legacy planner?

**Review by wuchong — COMMENTED** [review]
Thanks for the great work. I just roughly went through the code and left some comments.

File: flink-table/flink-table-planner-blink/pom.xml (hunk 1 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/pom.xml:194]: This is a required dependency now, the calcite code can not compile without this jar.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala (hunk 4 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala:71]: Yes, it can be removed.

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/schema/Statistic.java (hunk 5 above)
Comment (danny0405) [reply flink-table/flink-table-planner/src/main/java/org/apache/calcite/schema/Statistic.java:35]: Because the default value changes to null, the legacy planner does not have custom `Statistic` so there is no good way to override the default value, e.g. (the `Statistics.UNKNOWN`).

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java (hunk 3 above)
Comment (wuchong) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:997]: This is missed to update?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdPopulationSizeTest.scala [hunk 6]
```diff
@@ -294,28 +306,28 @@ class FlinkRelMdPopulationSizeTest extends FlinkRelMdHandlerTestBase {
...
       mq.getPopulationSize(logicalRightJoinOnLHSUniqueKeys, ImmutableBitSet.of(1, 5)), 1e-2)
-    assertEquals(19996088.14,
+    assertEquals(1.9996069026214965E7,
       mq.getPopulationSize(logicalRightJoinOnLHSUniqueKeys, ImmutableBitSet.of(0, 6)), 1e-2)
 
     assertEquals(1.0, mq.getPopulationSize(logicalFullJoinWithoutEquiCond, ImmutableBitSet.of()))
     assertEquals(2.0E7, mq.getPopulationSize(logicalFullJoinWithoutEquiCond, ImmutableBitSet.of(0)))
     assertEquals(8.0E8, mq.getPopulationSize(logicalFullJoinWithoutEquiCond, ImmutableBitSet.of(1)))
-    assertEquals(6.295509444597865E15,
+    assertEquals(8.0E15,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdPopulationSizeTest.scala:330]: is this within our expectations?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoLegacyTableSourceScanRuleTest.scala [hunk 7]
```diff
@@ -111,6 +111,9 @@ class PushProjectIntoLegacyTableSourceScanRuleTest extends TableTestBase {
 
   @Test
   def testProjectWithoutInputRef(): Unit = {
+    // Regression by: CALCITE-4220,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoLegacyTableSourceScanRuleTest.scala:114]: we can implement a rule to re-add a projection node for one phase aggregation

**Review by godfreyhe — CHANGES_REQUESTED** [review]
Thanks for the huge contribution, I left some comments.

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/schema/Statistic.java (hunk 5 above)
Comment (danny0405) [reply flink-table/flink-table-planner/src/main/java/org/apache/calcite/schema/Statistic.java:35]: Yes, but because the legacy planner use the `RelOptTableImpl` as the `SqlValidatorTable` during sql-to-rel conversion, and there is no common base preparing table like `FlinkPreparingTableBase` in blink planner, we need to override each `AbstractTable` sub-class for the legacy planner.

Here it is `QueryOperationCatalogViewTable` that causes the problem.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/AggSqlFunction.scala [hunk 8]
```diff
@@ -165,18 +165,28 @@ object AggSqlFunction {
...
     val methods = checkAndExtractMethods(aggregateFunction, "accumulate")
 
     /**
       * Operand type checker based on [[AggregateFunction]] given information.
       */
-    new SqlOperandTypeChecker {
+    new SqlOperandMetadata {
+      override def paramNames(): util.List[String] = {
+        // Does not support named parameters.
+        Collections.emptyList()
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/AggSqlFunction.scala:182]: What's the difference ?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableSqlFunction.scala [hunk 9]
```diff
@@ -40,12 +44,19 @@ class TableSqlFunction(
...
-    null,
+    createEvalOperandMetadata(name, tableFunction),
     functionImpl) {
 
+  override def getRowTypeInference: SqlReturnTypeInference = new SqlReturnTypeInference {
+    override def inferReturnType(opBinding: SqlOperatorBinding): RelDataType = {
+      // The arguments should never be used.
+      functionImpl.getRowType(opBinding.getTypeFactory, Collections.emptyList())
+    }
+  }
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableSqlFunction.scala:58]: Yes, all the sql table function needs to implement `getRowTypeInference` now.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 10]
```diff
@@ -372,6 +373,9 @@ object UserDefinedFunctionUtils {
 
       override def getConsistency: Consistency = Consistency.NONE
 
+      override def paramTypes(typeFactory: RelDataTypeFactory): util.List[RelDataType] = null
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:376]: It is never used, if there is NPE, it should be a bug.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/table/CalcTest.scala [hunk 11]
```diff
@@ -114,11 +114,13 @@ class CalcTest extends TableTestBase {
     val resultTable = sourceTable.select('a, 'b, 'c)
       .where((1 to 30).map($"b" === _).reduce((ex1, ex2) => ex1 || ex2) && ($"c" === "xx"))
 
+    val operands = "Sarg[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, " +
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/table/CalcTest.scala:117]: They are equivalent, we can use the `RexUtil.expandSearch` to convert it to `=(c, 'xx')`.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala [hunk 12]
```diff
@@ -2679,15 +2679,15 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
       "1017-11-29 22:58:58.998")
 
     val QUARTER = Seq(
-      "2018-03-01 22:58:58.998",
-      "2018-08-31 22:58:58.998",
+      "2018-02-28 22:58:58.998",
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala:2682]: See `CALCITE-3881` and `FLINK-16823`.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala [hunk 13]
```diff
@@ -204,20 +204,16 @@ class TimeIndicatorConversionTest extends TableTestBase {
...
-      unaryNode(
-        "DataStreamCalc",
         streamTableNode(t),
-        term("select", "rowtime")
+        term("all", "true"),
+        term("union all", "rowtime, long, int")
       ),
-      term("all", "true"),
-      term("union all", "rowtime")
+      term("select", "rowtime")
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala:216]: I guess CALCITE-3807 causes the problem.

[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala:216]: After some debug, i found that the 2 plan has the same cost:
image

[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala:216]: I would fire a issue to track this ~

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecNestedLoopJoin.scala [hunk 14]
```diff
@@ -94,7 +94,13 @@ class BatchExecNestedLoopJoin(
...
     val cpuCost = leftRowCnt * rightRowCnt
     val costFactory = planner.getCostFactory.asInstanceOf[FlinkCostFactory]
-    costFactory.makeCost(mq.getRowCount(this), cpuCost, 0, 0, memoryCost)
+    val cost = costFactory.makeCost(mq.getRowCount(this), cpuCost, 0, 0, memoryCost)
+    if (singleRowJoin) {
+      // Make single row join more preferable than non-single row join.
+      cost.multiplyBy(0.99)
+    } else {
+      cost
+    }
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecNestedLoopJoin.scala:103]: Without this change, the single row join plan can not be picked.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 15]
```diff
@@ -70,9 +70,29 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
   private def getTableUniqueKeys(
       tableSource: TableSource[_],
       relOptTable: RelOptTable): JSet[ImmutableBitSet] = {
-    // TODO get uniqueKeys from TableSchema of TableSource
 
     relOptTable match {
+      case sourceTable: TableSourceTable =>
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:75]: Because this pr is huge, i would give tests in another PR.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/SelectivityEstimator.scala [hunk 16]
```diff
@@ -187,8 +188,8 @@ class SelectivityEstimator(rel: RelNode, mq: FlinkRelMetadataQuery)
       case IS_NOT_NULL =>
         estimateIsNotNull(operands.head)
 
-      case IN =>
-        estimateIn(operands.head, operands.slice(1, operands.size))
+      case SEARCH =>
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/SelectivityEstimator.scala:191]: Yes, since 1.26. the `IN` constant rex call is not allowed, only in sub-query.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala [hunk 17]
```diff
@@ -343,6 +344,54 @@ object FlinkRexUtil {
...
+        null
+      }
+    }
+    node.accept(visitor)
+    false
+  } catch {
+    case e: Util.FoundOne =>
+      Util.swallow(e, null)
+      true
+  }
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:368]: It is used by the following PR, i can remove it now ~

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java [hunk 18]
```diff
@@ -768,7 +768,8 @@ public void testRowTableFunction() throws Exception {
 
 		tEnv().createTemporarySystemFunction("RowTableFunction", RowTableFunction.class);
 		tEnv().executeSql(
-				"INSERT INTO SinkTable SELECT t.s, t.sa FROM SourceTable, LATERAL TABLE(RowTableFunction(s)) t")
+				"INSERT INTO SinkTable SELECT t.s, t.sa FROM SourceTable source, "
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:771]: Yes, the same column name `s` has ambiguity now.

File: flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/join/SemiAntiJoinTest.xml [hunk 19]
```diff
@@ -516,28 +516,28 @@ LogicalFilter(condition=[<>($cor0.b, $1)])
       <![CDATA[
 Join(joinType=[LeftAntiJoin], where=[<>(b, e)], select=[a, b, c], leftInputSpec=[NoUniqueKey], rightInputSpec=[NoUniqueKey])
 :- Exchange(distribution=[single])
-:  +- Join(joinType=[LeftSemiJoin], where=[$f0], select=[a, b, c], leftInputSpec=[NoUniqueKey], rightInputSpec=[NoUniqueKey])
-:     :- Exchange(distribution=[single])
-:     :  +- Join(joinType=[LeftAntiJoin], where=[AND(OR(=(b, i), IS NULL(b), IS NULL(i)), =(c, k))], select=[a, b, c], leftInputSpec=[NoUniqueKey], rightInputSpec=[NoUniqueKey])
-:     :     :- Exchange(distribution=[hash[c]])
+:  +- Join(joinType=[LeftAntiJoin], where=[AND(OR(IS NULL(b), IS NULL(i), =(b, i)), =(c, k))], select=[a, b, c], leftInputSpec=[NoUniqueKey], rightInputSpec=[NoUniqueKey])
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/join/SemiAntiJoinTest.xml:519]: Actually i'm not sure.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdPopulationSizeTest.scala (hunk 6 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdPopulationSizeTest.scala:330]: I'm not sure, but the only change is the algorithm of `NDV`.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/TimeIndicatorConversionTest.scala:216]: See FLINK-19668

File: flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/table/SetOperatorsTest.xml [hunk 20]
```diff
@@ -100,10 +100,9 @@ LogicalProject(b=[$1], c=[$2])
     </Resource>
     <Resource name="planAfter">
       <![CDATA[
-Union(all=[true], union=[b, c])
-:- Calc(select=[b, c])
-:  +- LegacyTableSourceScan(table=[[default_catalog, default_database, left, source: [TestTableSource(a, b, c)]]], fields=[a, b, c])
-+- Calc(select=[b, c])
+Calc(select=[b, c])
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/table/SetOperatorsTest.xml:103]: Because the cost is the same, tracked by FLINK-19668.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/AggSqlFunction.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/AggSqlFunction.scala:182]: we had introduced `java.util`.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableSqlFunction.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableSqlFunction.scala:58]: `TableSqlFunction` extends from `SqlUserDefinedTableFunction`, and `SqlUserDefinedTableFunction` has provided the default implementation

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:376]: throw exception directly

**Comment by godfreyhe** [comment]
Thanks for the update and detailed explanation,  LGTM overall except a few minor comments

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/AggSqlFunction.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/AggSqlFunction.scala:182]: Thanks, `java.util` import removed.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableSqlFunction.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableSqlFunction.scala:58]: Yes, it can be removed.

File: flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/batch/sql/agg/WindowAggregateTest.xml [hunk 21]
```diff
@@ -699,10 +699,10 @@ LogicalProject(s=[$1], a=[$2], wStart=[TUMBLE_START($0)])
     </Resource>
     <Resource name="planAfter">
       <![CDATA[
-Calc(select=[CAST(CASE(=($f1, 0), null:INTEGER, s)) AS s, CAST(CAST(/(CASE(=($f1, 0), null:INTEGER, s), $f1))) AS a, w$start AS wStart])
-+- HashWindowAggregate(window=[TumblingGroupWindow('w$, b, 900000)], properties=[w$start, w$end, w$rowtime], select=[Final_$SUM0(sum$0) AS s, Final_COUNT(count1$1) AS $f1])
+Calc(select=[CAST(s) AS s, CAST(a) AS a, w$start AS wStart])
++- HashWindowAggregate(window=[TumblingGroupWindow('w$, b, 900000)], properties=[w$start, w$end, w$rowtime], select=[Final_SUM(sum$0) AS s, Final_AVG(sum$1, count$2) AS a])
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/batch/sql/agg/WindowAggregateTest.xml:703]: image
The cost are the same with the old plan, same cause as the `Union` project push case.

[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/batch/sql/agg/WindowAggregateTest.xml:703]: Tracked by `FLINK-19668`.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala [hunk 22]
```diff
@@ -58,6 +61,18 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
...
+      rel: TableFunctionScan,
+      mq: RelMetadataQuery,
+      ignoreNulls: Boolean): JSet[ImmutableBitSet] = {
+    if (rel.getInputs.size() == 1
+        && rel.getCall.asInstanceOf[RexCall].getOperator.isInstanceOf[SqlWindowTableFunction]) {
+      mq.getUniqueKeys(rel.getInput(0), ignoreNulls)
+    } else {
+      null
+    }
+  }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:74]: is this change necessary ?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdPopulationSizeTest.scala [hunk 23]
```diff
@@ -66,19 +66,25 @@ class FlinkRelMdPopulationSizeTest extends FlinkRelMdHandlerTestBase {
     assertEquals(1.0, mq.getPopulationSize(logicalProject, ImmutableBitSet.of()))
     assertEquals(50.0, mq.getPopulationSize(logicalProject, ImmutableBitSet.of(0)))
     assertEquals(48.0, mq.getPopulationSize(logicalProject, ImmutableBitSet.of(1)))
-    assertEquals(16.22, mq.getPopulationSize(logicalProject, ImmutableBitSet.of(2)), 1e-2)
-    assertEquals(6.98, mq.getPopulationSize(logicalProject, ImmutableBitSet.of(3)), 1e-2)
-    assertEquals(20.09, mq.getPopulationSize(logicalProject, ImmutableBitSet.of(4)), 1e-2)
-    assertEquals(20.09, mq.getPopulationSize(logicalProject, ImmutableBitSet.of(5)), 1e-2)
+    assertEquals(16.43531528030365,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdPopulationSizeTest.scala:69]: this also need to update

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/ByteHashSet.java [hunk 24]
```diff
@@ -26,7 +26,7 @@
 
 	protected boolean[] used;
 
-	public ByteHashSet() {
+	public ByteHashSet(final int expected) {
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/ByteHashSet.java:29]: unnecessary change

Comment (danny0405) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/collections/ByteHashSet.java:29]: This is needed because it is a bug.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala (hunk 22 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala:74]: No, i can remove it.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/TableSqlFunction.scala [hunk 25]
```diff
@@ -82,9 +83,16 @@ class TableSqlFunction(
 
   override def toString: String = displayName
 
-  override def getRowType(
+  override def getRowTypeInference: SqlReturnTypeInference = new SqlReturnTypeInference {
+    override def inferReturnType(opBinding: SqlOperatorBinding): RelDataType = {
+      val arguments = convertArguments(opBinding, functionImpl, getNameAsId)
+      getRowType(opBinding.getTypeFactory, arguments)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/functions/utils/TableSqlFunction.scala:91]: It is needed because `getRowType` needs to be overridden.

**Review by godfreyhe — APPROVED** [review]
thanks for the update, LGTM, +1 to merge
